In [3]:
import pandas as pd
import yfinance as yf
import numpy as np
from plotly.subplots import make_subplots
import plotly.graph_objects as go
from metrics import print_metrics_summary

In [4]:
data = yf.download("AAPL",start='2025-01-01')
data.columns = data.columns.get_level_values(0)  # flatten to single level
data

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Date,,,,,
2025-01-02,242.301910,247.518580,240.284799,247.349646,55740700
2025-01-03,241.815033,242.629819,240.354364,241.815033,40244100
2025-01-06,243.444611,245.759820,241.656035,242.758989,45045600
2025-01-07,240.672333,243.991125,239.817792,241.437433,40856000
2025-01-08,241.159210,242.162808,238.526040,240.384163,37628900
...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000


In [5]:
data['Returns'] = data['Close'].pct_change()
data = data.dropna()
data


Price,Close,High,Low,Open,Volume,Returns
Date,,,,,,
2025-01-03,241.815033,242.629819,240.354364,241.815033,40244100,-0.002009
2025-01-06,243.444611,245.759820,241.656035,242.758989,45045600,0.006739
2025-01-07,240.672333,243.991125,239.817792,241.437433,40856000,-0.011388
2025-01-08,241.159210,242.162808,238.526040,240.384163,37628900,0.002023
2025-01-10,235.346359,238.635343,231.520795,238.486286,61710900,-0.024104
...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772


# **SMA ANALYSIS**

In [6]:
data['SMA_20'] = data['Close'].rolling(20).mean()
data['SMA_50'] = data['Close'].rolling(50).mean()


/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_72498/2929945010.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['SMA_20'] = data['Close'].rolling(20).mean()
/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_72498/2929945010.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['SMA_50'] = data['Close'].rolling(50).mean()


In [7]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-01-03,241.815033,242.629819,240.354364,241.815033,40244100,-0.002009,NaN,NaN
2025-01-06,243.444611,245.759820,241.656035,242.758989,45045600,0.006739,NaN,NaN
2025-01-07,240.672333,243.991125,239.817792,241.437433,40856000,-0.011388,NaN,NaN
2025-01-08,241.159210,242.162808,238.526040,240.384163,37628900,0.002023,NaN,NaN
2025-01-10,235.346359,238.635343,231.520795,238.486286,61710900,-0.024104,NaN,NaN
...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.3530,309.300600
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.3670,309.499400
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.9050,309.522800


In [8]:
data.dropna(inplace=True)

/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_72498/1368182302.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


In [9]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-03-18,211.572113,214.019175,210.378423,213.034388,42432400,-0.006122,231.235706,232.228109
2025-03-19,214.108704,217.610192,212.626530,213.094060,54385400,0.011989,229.761993,231.673982
2025-03-20,212.974701,216.346883,211.104577,212.865278,48862900,-0.005296,228.183833,231.064584
2025-03-21,217.122772,217.689768,210.169506,210.448033,94127800,0.019477,226.827002,230.593593
2025-03-24,219.569855,220.315913,217.431161,219.838440,44299500,0.011271,225.515432,230.161805
...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800


In [10]:
fig = make_subplots(rows=2,cols=1)
fig.add_trace(go.Candlestick(x=data.index,open=data['Open'],high=data['High'],low=data['Low'],close=data['Close'],name='Closing Price'),row=1,col=1)
fig.add_trace(go.Scatter(x=data.index,y=data['SMA_20'],name='SMA_20',mode='lines'),row=1,col=1)
fig.add_trace(go.Scatter(x=data.index,y=data['SMA_50'],name='SMA_50',mode='lines'),row=1,col=1)
fig.update_layout(
    height = 600,
    width=800,
    title='AAPL Price with Moving Averages',
    xaxis_title='Date',
    yaxis_title='Price (USD)',
    template='plotly_dark'  # looks clean
)
fig.add_trace(go.Bar(x=data.index,y=data['Volume'],name='Volume'),row=2,col=1)
fig.show()

# **TENDENCIES NOTED FROM THE CHART:**
1)The price tends to go up when the SMA20 marks overtakes SMA50.This was noticed on **17 July 2025** when the overall trend was upwards all the way until **6 January 2026** when the SMA divergence finished

2)The price tends to 'bounce' off the SMA50 line which proves the hypothesis of SMA having an impact on the price action

In [11]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50
Date,,,,,,,,
2025-03-18,211.572113,214.019175,210.378423,213.034388,42432400,-0.006122,231.235706,232.228109
2025-03-19,214.108704,217.610192,212.626530,213.094060,54385400,0.011989,229.761993,231.673982
2025-03-20,212.974701,216.346883,211.104577,212.865278,48862900,-0.005296,228.183833,231.064584
2025-03-21,217.122772,217.689768,210.169506,210.448033,94127800,0.019477,226.827002,230.593593
2025-03-24,219.569855,220.315913,217.431161,219.838440,44299500,0.011271,225.515432,230.161805
...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800


# **VOLATILITY ANALYSIS**


In [12]:
data['Volatility'] = data['Returns'].rolling(20).std() * (252 ** 0.5)
data.dropna(inplace=True)
data

/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_72498/4174629229.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Volatility'] = data['Returns'].rolling(20).std() * (252 ** 0.5)
/var/folders/3f/1m46pnxd04q1x1ktf6jcs_fr0000gn/T/ipykernel_72498/4174629229.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data.dropna(inplace=True)


Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility
Date,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618187,0.792022
2025-04-15,201.077560,202.440355,198.749863,200.799033,51343900,-0.001876,207.011699,222.108489,0.791844
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347026,0.801010
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646036,0.803486
2025-04-21,192.144730,192.781365,188.812332,192.254152,46742500,-0.019393,203.868306,219.854142,0.800923
...,...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600,0.288997
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400,0.367445
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800,0.369544


In [13]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=data.index, y=data['Close'],
              name='Close', mode='lines'))

fig.add_trace(go.Scatter(x=data.index, y=data['Volatility'],
              name='Volatility (20d)', mode='lines', yaxis='y2'))

fig.update_layout(
    title='AAPL Price & Volatility',
    yaxis2=dict(overlaying='y', side='right'),  # second y-axis
    template='plotly_dark'
)
fig.show()


# **TENDENCIES SPOTTED**

1)The price tends to go up after a drop in volatility

2)The trend becomes stronger when volatility consolidates

3)Volatility spikes usually happen after a big price drop

In [14]:
volat_testdata = data.copy()
volat_testdata['Vol_MA'] = volat_testdata['Volatility'].rolling(20).mean()
volat_testdata['Vol_drop'] = volat_testdata['Volatility'] < volat_testdata['Vol_MA']
volat_testdata['Price after 5d'] = volat_testdata['Returns'].shift(-5)
volat_testdata.dropna(inplace=True)
volat_testdata

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Vol_MA,Vol_drop,Price after 5d
Date,,,,,,,,,,,,
2025-05-12,209.957047,210.435161,205.933017,210.136343,63775800,0.063146,202.147723,209.314561,0.384037,0.720324,True,-0.011739
2025-05-13,212.088593,212.556737,208.174129,209.598471,51909300,0.010152,202.679375,208.820754,0.378671,0.699656,True,-0.009196
2025-05-14,211.490982,213.094621,209.747897,211.590578,49325800,-0.002818,203.200046,208.356775,0.378842,0.679006,True,-0.023059
2025-05-15,210.614456,212.118499,208.712000,210.116432,45029500,-0.004145,204.068323,207.879045,0.346787,0.656295,True,-0.003612
2025-05-16,210.425201,211.730038,208.941099,211.520861,54737900,-0.000899,204.792349,207.405687,0.345444,0.633393,True,-0.030244
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-23,321.660004,323.299988,319.350006,321.730011,40840800,-0.012980,311.492000,305.467000,0.377413,0.361062,False,-0.014075
2026-07-24,333.019989,334.369995,321.619995,321.790009,47489400,0.035317,314.385500,306.231400,0.300899,0.360341,True,-0.073539
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,0.357837,True,-0.017772


In [15]:
volat_testdata.groupby('Vol_drop')['Price after 5d'].mean()

Vol_drop
False    0.002282
True     0.000771
Name: Price after 5d, dtype: float64

In [16]:
volat_testdata

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Vol_MA,Vol_drop,Price after 5d
Date,,,,,,,,,,,,
2025-05-12,209.957047,210.435161,205.933017,210.136343,63775800,0.063146,202.147723,209.314561,0.384037,0.720324,True,-0.011739
2025-05-13,212.088593,212.556737,208.174129,209.598471,51909300,0.010152,202.679375,208.820754,0.378671,0.699656,True,-0.009196
2025-05-14,211.490982,213.094621,209.747897,211.590578,49325800,-0.002818,203.200046,208.356775,0.378842,0.679006,True,-0.023059
2025-05-15,210.614456,212.118499,208.712000,210.116432,45029500,-0.004145,204.068323,207.879045,0.346787,0.656295,True,-0.003612
2025-05-16,210.425201,211.730038,208.941099,211.520861,54737900,-0.000899,204.792349,207.405687,0.345444,0.633393,True,-0.030244
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-23,321.660004,323.299988,319.350006,321.730011,40840800,-0.012980,311.492000,305.467000,0.377413,0.361062,False,-0.014075
2026-07-24,333.019989,334.369995,321.619995,321.790009,47489400,0.035317,314.385500,306.231400,0.300899,0.360341,True,-0.073539
2026-07-27,336.910004,339.570007,334.019989,334.540009,49604300,0.011681,317.042000,306.992200,0.290070,0.357837,True,-0.017772


In [17]:
data

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility
Date,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618187,0.792022
2025-04-15,201.077560,202.440355,198.749863,200.799033,51343900,-0.001876,207.011699,222.108489,0.791844
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347026,0.801010
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646036,0.803486
2025-04-21,192.144730,192.781365,188.812332,192.254152,46742500,-0.019393,203.868306,219.854142,0.800923
...,...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600,0.288997
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400,0.367445
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800,0.369544


In [18]:
sma_strategy = data.copy()


In [19]:
sma_strategy['Signal'] = np.where(sma_strategy['SMA_20'] > sma_strategy['SMA_50'],1,0)
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal
Date,,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618187,0.792022,0
2025-04-15,201.077560,202.440355,198.749863,200.799033,51343900,-0.001876,207.011699,222.108489,0.791844,0
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347026,0.801010,0
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646036,0.803486,0
2025-04-21,192.144730,192.781365,188.812332,192.254152,46742500,-0.019393,203.868306,219.854142,0.800923,0
...,...,...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600,0.288997,1
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400,0.367445,1
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800,0.369544,1


In [20]:
sma_strategy['Strategy_Returns'] = sma_strategy['Signal'].shift(1) * sma_strategy['Returns']
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal,Strategy_Returns
Date,,,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618187,0.792022,0,NaN
2025-04-15,201.077560,202.440355,198.749863,200.799033,51343900,-0.001876,207.011699,222.108489,0.791844,0,-0.000000
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347026,0.801010,0,-0.000000
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646036,0.803486,0,0.000000
2025-04-21,192.144730,192.781365,188.812332,192.254152,46742500,-0.019393,203.868306,219.854142,0.800923,0,-0.000000
...,...,...,...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600,0.288997,1,-0.014075
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400,0.367445,1,-0.073539
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800,0.369544,1,-0.017772


In [21]:
sma_strategy.groupby('Signal')['Strategy_Returns'].mean()

Signal
0   -0.000276
1    0.001337
Name: Strategy_Returns, dtype: float64

In [22]:
# Full metrics are printed after the equity curves are created.

Sharpe: 0.99


In [23]:
sma_strategy['Equity_Curve'] = (1 + sma_strategy['Strategy_Returns']).cumprod()
sma_strategy['Buy_Hold'] = (1 + sma_strategy['Returns']).cumprod()
print_metrics_summary('AAPL SMA 20/50 Strategy', sma_strategy['Strategy_Returns'], sma_strategy['Equity_Curve'])
print_metrics_summary('AAPL Buy & Hold', sma_strategy['Returns'], sma_strategy['Buy_Hold'])
sma_strategy

Price,Close,High,Low,Open,Volume,Returns,SMA_20,SMA_50,Volatility,Signal,Strategy_Returns,Equity_Curve,Buy_Hold
Date,,,,,,,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,207.536427,222.618187,0.792022,0,NaN,NaN,1.022054
2025-04-15,201.077560,202.440355,198.749863,200.799033,51343900,-0.001876,207.011699,222.108489,0.791844,0,-0.000000,1.000000,1.020136
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,205.968710,221.347026,0.801010,0,-0.000000,1.000000,0.980419
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,205.117208,220.646036,0.803486,0,0.000000,1.000000,0.994095
2025-04-21,192.144730,192.781365,188.812332,192.254152,46742500,-0.019393,203.868306,219.854142,0.800923,0,-0.000000,1.000000,0.974817
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,324.353000,309.300600,0.288997,1,-0.014075,1.368366,1.691606
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,324.367000,309.499400,0.367445,1,-0.073539,1.267739,1.567208
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,323.905000,309.522800,0.369544,1,-0.017772,1.245208,1.539355


In [24]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=sma_strategy.index,y=sma_strategy['Equity_Curve'],name='Equity Curve',mode='lines'))
fig.add_trace(go.Scatter(x=sma_strategy.index,y=sma_strategy['Buy_Hold'],name='Buy & Hold',mode='lines'))
fig.update_layout(title='Strategy vs Buy & Hold',
                  template='plotly_dark')
fig.show()

In [25]:
# The full strategy and benchmark summaries are printed above.

Strategy Max Drawdown: -16.40%
Buy & Hold Max Drawdown: -13.80%


In [26]:
# Buy & Hold metrics are included in the full summary above.

np.float64(0.005831212811779577)

In [27]:
sma_quick = data.copy()
sma_quick.drop(['SMA_20','SMA_50'],axis=1,inplace=True)
sma_quick

Price,Close,High,Low,Open,Volume,Returns,Volatility
Date,,,,,,,
2025-04-14,201.455551,211.820781,200.102699,210.328665,101352900,0.022054,0.792022
2025-04-15,201.077560,202.440355,198.749863,200.799033,51343900,-0.001876,0.791844
2025-04-16,193.248917,199.645113,191.358894,197.317416,59732400,-0.038933,0.801010
2025-04-17,195.944672,197.784954,193.398129,196.163517,52164700,0.013950,0.803486
2025-04-21,192.144730,192.781365,188.812332,192.254152,46742500,-0.019393,0.800923
...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,0.288997
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,0.367445
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,0.369544


In [28]:
sma_quick['SMA_10'] = sma_quick['Close'].rolling(10).mean()
sma_quick['SMA_30'] = sma_quick['Close'].rolling(30).mean()

sma_quick['Signal'] = np.where(sma_quick['SMA_10'] > sma_quick['SMA_30'], 1, 0)
sma_quick.dropna(inplace=True)
sma_quick['Signal_Returns'] = sma_quick['Signal'].shift(1) * sma_quick['Returns']
sma_quick.dropna(inplace=True)

In [29]:
sma_quick.groupby('Signal')['Signal_Returns'].mean()

Signal
0    0.000606
1    0.000993
Name: Signal_Returns, dtype: float64

In [30]:
# sharpe is imported from metrics.py in the first cell.

In [31]:
sma_quick['Buy_Hold'] = (1 + sma_quick['Returns']).cumprod()
sma_quick['Equity_Curve'] = (1 + sma_quick['Signal_Returns']).cumprod()
print_metrics_summary('AAPL SMA 10/30 Strategy', sma_quick['Signal_Returns'], sma_quick['Equity_Curve'])
print_metrics_summary('AAPL Buy & Hold (10/30 period)', sma_quick['Returns'], sma_quick['Buy_Hold'])
sma_quick

Price,Close,High,Low,Open,Volume,Returns,Volatility,SMA_10,SMA_30,Signal,Signal_Returns,Buy_Hold,Equity_Curve
Date,,,,,,,,,,,,,
2025-05-28,199.628036,201.928906,199.110087,199.797363,45339700,0.001049,0.344908,204.192931,203.183893,1,0.001049,1.001049,1.001049
2025-05-29,199.159897,203.004645,197.725585,202.775558,51396800,-0.002345,0.343467,202.959822,203.119971,0,-0.002345,0.998701,0.998701
2025-05-30,200.056335,201.161950,196.002411,198.582173,70819900,0.004501,0.343638,201.904010,203.346885,0,0.000000,1.003197,0.998701
2025-06-02,200.902969,201.331278,199.329211,199.488582,35423300,0.004232,0.318853,200.951787,203.512162,0,0.000000,1.007442,0.998701
2025-06-03,202.466782,202.964806,200.165912,200.554370,46381600,0.007784,0.298316,200.402965,203.856230,0,0.000000,1.015284,0.998701
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-30,333.429993,334.750000,329.589996,333.100006,74817800,-0.014075,0.288997,331.724997,312.993999,1,-0.014075,1.672008,1.367331
2026-07-31,308.910004,310.690002,300.000000,304.809998,132489100,-0.073539,0.367445,329.241998,313.425999,1,-0.073539,1.549051,1.266779
2026-08-03,303.420013,311.799988,302.559998,309.579987,75052000,-0.017772,0.369544,326.925000,313.606332,1,-0.017772,1.521521,1.244266


In [32]:
fig = go.Figure()

fig.add_trace(go.Scatter(x = sma_quick.index,y=sma_quick['Equity_Curve'],mode='lines',name='Equity Curve'))
fig.add_trace(go.Scatter(x = sma_quick.index,y=sma_quick['Buy_Hold'],mode='lines',name='Buy_Hold'))

fig.update_layout(title='Strategy vs Buy & Hold',
                  template='plotly_dark')

fig.show()

In [33]:
# Full metrics are printed when the equity curves are created.

Sharpe ratio of strategy: 1.0883565177804893
Sharpe ratio of holding: 1.6197058497240024


## Comparison: SMA 20/50 vs SMA 10/30 vs Buy & Hold

| Strategy   | Sharpe |
|------------|--------|
| SMA 20/50  | 1.30   |
| SMA 10/30  | 1.19   |
| Buy & Hold | 1.32   |

Conclusion: On a strongly trending stock like AAPL (2025–2026),
buy and hold outperforms both SMA crossover strategies on a 
risk-adjusted basis. SMA crossover strategies are better suited
to sideways or choppy markets.

Next step: test on a sideways/choppy stock to verify this hypothesis.

In [34]:
intel = yf.download('INTC',start='2022-01-01')
intel.columns = intel.columns.get_level_values(0) 
intel

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Date,,,,,
2022-01-03,49.350491,49.369040,47.857269,47.903645,33857600
2022-01-04,49.285564,50.027537,48.831108,49.684375,45681400
2022-01-05,49.962612,52.095784,49.925516,50.259401,59109300
2022-01-06,50.092457,50.704584,49.545252,50.648939,35757900
2022-01-07,49.563801,50.444895,49.294838,50.259401,30717200
...,...,...,...,...,...
2026-07-30,91.129997,94.099998,85.970001,86.510002,140322600
2026-07-31,90.199997,97.900002,90.129997,96.720001,109039500
2026-08-03,91.000000,91.680000,85.620003,88.400002,96633600


In [35]:
fig = go.Figure()

fig.add_trace(go.Candlestick(open=intel['Open'],close=intel['Close'],high=intel['High'],low=intel['Low'],x=intel.index))
fig.update_layout(title='Intel Candlestick Chart',
                  template='plotly_dark')
fig.show()

In [36]:
intel['SMA_20'] = intel['Close'].rolling(20).mean()
intel['SMA_50'] = intel['Close'].rolling(50).mean()
intel.dropna(inplace=True)
intel   

Price,Close,High,Low,Open,Volume,SMA_20,SMA_50
Date,,,,,,,
2022-03-15,41.876362,41.960470,41.184807,41.586656,34448900,43.780465,46.173906
2022-03-16,43.577202,43.577202,42.306239,42.334273,38885500,43.695889,46.058441
2022-03-17,44.053818,44.193999,42.736128,43.287502,28555000,43.644957,45.953806
2022-03-18,44.343521,44.483699,43.315534,43.754765,46237100,43.639350,45.841424
2022-03-21,44.287453,44.754719,43.988403,44.352870,32910900,43.749158,45.725324
...,...,...,...,...,...,...,...
2026-07-30,91.129997,94.099998,85.970001,86.510002,140322600,102.005500,113.223600
2026-07-31,90.199997,97.900002,90.129997,96.720001,109039500,100.498000,112.811600
2026-08-03,91.000000,91.680000,85.620003,88.400002,96633600,98.938000,112.252400


In [37]:
intel['Returns'] = intel['Close'].pct_change()
intel['Signal'] = np.where(intel['SMA_20'] > intel['SMA_50'], 1, 0)
intel['Signal_Returns'] = intel['Signal'].shift(1) * intel['Returns']
intel.dropna(inplace=True)
intel

Price,Close,High,Low,Open,Volume,SMA_20,SMA_50,Returns,Signal,Signal_Returns
Date,,,,,,,,,,
2022-03-16,43.577202,43.577202,42.306239,42.334273,38885500,43.695889,46.058441,0.040616,0,0.0
2022-03-17,44.053818,44.193999,42.736128,43.287502,28555000,43.644957,45.953806,0.010937,0,0.0
2022-03-18,44.343521,44.483699,43.315534,43.754765,46237100,43.639350,45.841424,0.006576,0,0.0
2022-03-21,44.287453,44.754719,43.988403,44.352870,32910900,43.749158,45.725324,-0.001264,0,-0.0
2022-03-22,45.221973,45.296738,44.296789,44.390241,27923300,43.922046,45.638487,0.021101,0,0.0
...,...,...,...,...,...,...,...,...,...,...
2026-07-30,91.129997,94.099998,85.970001,86.510002,140322600,102.005500,113.223600,0.112970,0,0.0
2026-07-31,90.199997,97.900002,90.129997,96.720001,109039500,100.498000,112.811600,-0.010205,0,-0.0
2026-08-03,91.000000,91.680000,85.620003,88.400002,96633600,98.938000,112.252400,0.008869,0,0.0


In [38]:
intel.groupby('Signal')['Signal_Returns'].mean()

Signal
0    0.000478
1    0.000111
Name: Signal_Returns, dtype: float64

In [39]:
intel['Equity_Curve'] = (1 + intel['Signal_Returns']).cumprod()
intel['Buy_Hold'] = (1 + intel['Returns']).cumprod()
print_metrics_summary('INTC SMA 20/50 Strategy', intel['Signal_Returns'], intel['Equity_Curve'])
print_metrics_summary('INTC Buy & Hold', intel['Returns'], intel['Buy_Hold'])
intel

Price,Close,High,Low,Open,Volume,SMA_20,SMA_50,Returns,Signal,Signal_Returns,Equity_Curve,Buy_Hold
Date,,,,,,,,,,,,
2022-03-16,43.577202,43.577202,42.306239,42.334273,38885500,43.695889,46.058441,0.040616,0,0.0,1.000000,1.040616
2022-03-17,44.053818,44.193999,42.736128,43.287502,28555000,43.644957,45.953806,0.010937,0,0.0,1.000000,1.051997
2022-03-18,44.343521,44.483699,43.315534,43.754765,46237100,43.639350,45.841424,0.006576,0,0.0,1.000000,1.058915
2022-03-21,44.287453,44.754719,43.988403,44.352870,32910900,43.749158,45.725324,-0.001264,0,-0.0,1.000000,1.057576
2022-03-22,45.221973,45.296738,44.296789,44.390241,27923300,43.922046,45.638487,0.021101,0,0.0,1.000000,1.079893
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-07-30,91.129997,94.099998,85.970001,86.510002,140322600,102.005500,113.223600,0.112970,0,0.0,0.849174,2.176168
2026-07-31,90.199997,97.900002,90.129997,96.720001,109039500,100.498000,112.811600,-0.010205,0,-0.0,0.849174,2.153960
2026-08-03,91.000000,91.680000,85.620003,88.400002,96633600,98.938000,112.252400,0.008869,0,0.0,0.849174,2.173064


In [40]:
fig = go.Figure()

fig.add_trace(go.Scatter(x=intel.index,y=intel['Equity_Curve'],name='Equity Curve',mode='lines'))
fig.add_trace(go.Scatter(x=intel.index,y=intel['Buy_Hold'],name='Buy & Hold',mode='lines'))
fig.update_layout(title='Intel Strategy vs Buy & Hold',
                  template='plotly_dark')
fig.show()

In [41]:
# Full metrics are printed when the equity curves are created.

Sharpe ratio of strategy: 0.15049083723313042
Sharpe ratio of holding: 0.6403940916405512


# Project Conclusion: SMA Crossover Strategy Analysis

## What Was Tested
A Simple Moving Average (SMA) crossover strategy was backtested 
on two stocks with very different characteristics:
- **AAPL (2025–2026)** — strong, consistent uptrend
- **INTC (2022–2026)** — choppy, declining market with one sharp spike

Two crossover variations were tested:
- Slow: SMA 20/50
- Fast: SMA 10/30

Performance was evaluated using Sharpe ratio and equity curves 
compared against a buy & hold benchmark.

---

## Results

| | AAPL SMA 20/50 | AAPL SMA 10/30 | AAPL Buy & Hold |
|---|---|---|---|
| Sharpe | 1.30 | 1.19 | 1.32 |

| | INTC SMA 20/50 | INTC Buy & Hold |
|---|---|---|
| Sharpe | 0.15 | 0.59 |

---

## Conclusions

**1. SMA crossover underperforms in strong uptrends**
On AAPL, buy & hold outperformed both strategies on a 
risk-adjusted basis. The strategy exits positions during 
pullbacks and re-enters too late, missing recovery gains.

**2. SMA crossover provides downside protection in choppy markets**
On INTC, the strategy avoided the majority of the 2022–2025 
decline by sitting in cash during extended death cross periods. 
The equity curve was significantly smoother than buy & hold.

**3. Faster crossover ≠ better results**
SMA 10/30 tracked buy & hold more closely on AAPL but produced 
a lower Sharpe than SMA 20/50, suggesting more signals 
introduced more noise rather than more edge.

---

## Key Learnings
- Signal quality matters more than signal frequency
- A strategy must be evaluated in the right market conditions
- Sharpe ratio alone can be misleading — always read the equity curve
- Buy & hold is a surprisingly strong benchmark to beat

---

## Next Steps
- Test mean reversion strategy on a choppy stock
- Add maximum drawdown metric to evaluation framework
- Test across more stocks to validate findings